flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPool.java

```diff
@@ -266,104 +279,367 @@ public void disconnectResourceManager() {
+			SlotSharingManager multiTaskSlotManager,
+			ResourceProfile resourceProfile,
+			Collection<TaskManagerLocation> locationPreferences,
+			boolean allowQueuedScheduling) throws NoResourceAvailableException {
+		final SlotRequestId coLocationSlotRequestId = coLocationConstraint.getSlotRequestId();
+
+		if (coLocationSlotRequestId != null) {
+			// we have a slot assigned --> try to retrieve it
+			final SlotSharingManager.TaskSlot taskSlot = multiTaskSlotManager.getTaskSlot(coLocationSlotRequestId);
+
+			if (taskSlot != null) {
+				Preconditions.checkState(taskSlot instanceof SlotSharingManager.MultiTaskSlot);
+				return SlotSharingManager.MultiTaskSlotLocality.of(((SlotSharingManager.MultiTaskSlot) taskSlot), Locality.LOCAL);
+			} else {
+				// the slot may have been cancelled in the mean time
+				coLocationConstraint.setSlotRequestId(null);
+			}
+		}
 
-		if (pendingRequest != null) {
-			failPendingRequest(pendingRequest, new CancellationException("Allocation with request id" + requestId + " cancelled."));
+		final Collection<TaskManagerLocation> actualLocationPreferences;
+
+		if (coLocationConstraint.isAssigned()) {
+			actualLocationPreferences = Collections.singleton(coLocationConstraint.getLocation());
 		} else {
-			final Slot slot = allocatedSlots.get(requestId);
+			actualLocationPreferences = locationPreferences;
+		}
+
+		// get a new multi task slot
+		final SlotSharingManager.MultiTaskSlotLocality multiTaskSlotLocality = allocateMultiTaskSlot(
+			coLocationConstraint.getGroupId(), multiTaskSlotManager,
+			resourceProfile,
+			actualLocationPreferences,
+			allowQueuedScheduling);
+
+		// check whether we fulfill the co-location constraint
+		if (coLocationConstraint.isAssigned() && multiTaskSlotLocality.getLocality() != Locality.LOCAL) {
+			multiTaskSlotLocality.getMultiTaskSlot().release(
+				new FlinkException("Multi task slot is not local and, thus, does not fulfill the co-location constraint."));
 
-			if (slot != null) {
-				LOG.info("Returning allocated slot {} because the corresponding allocation request {} was cancelled.", slot, requestId);
-				if (slot.markCancelled()) {
-					internalReturnAllocatedSlot(slot);
+			throw new NoResourceAvailableException("Could not allocate a local multi task slot for the " +
+				"co location constraint " + coLocationConstraint + '.');
+		}
+
+		final SlotRequestId slotRequestId = new SlotRequestId();
+		final SlotSharingManager.MultiTaskSlot coLocationSlot = multiTaskSlotLocality.getMultiTaskSlot().allocateMultiTaskSlot(
+			slotRequestId,
+			coLocationConstraint.getGroupId());
+
+		// mark the requested slot as co-located slot for other co-located tasks
+		coLocationConstraint.setSlotRequestId(slotRequestId);
+
+		// lock the co-location constraint once we have obtained the allocated slot
+		coLocationSlot.getSlotContextFuture().whenComplete(
+			(SlotContext slotContext, Throwable throwable) -> {
+				if (throwable == null) {
```

Comment:
Is it acceptable to swallow the `Throwable` here?

Reply:
I will log the throwable here.